src/workflows/zip/ZipWorkflowManager.ts

```diff
@@ -0,0 +1,182 @@
+import { ReadStream, createReadStream, createWriteStream } from 'fs';
+import { Readable } from 'stream';
+
+import { logger } from '@user-office-software/duo-logger';
+import archiver from 'archiver';
+
+import AttachmentFactory, { AttachmentFactoryMeta } from './AttachmentFactory';
+import { failSafeDeleteFiles, generateTmpPath } from '../../util/fileSystem';
+import { ResponseHeader, WorkflowManager } from '../WorkflowManager';
+
+type Constructable<T> = {
+  new (entityId: number): T;
+};
+type FactoryMeta<TZipFactoryMeta extends AttachmentFactoryMeta> = {
+  meta: TZipFactoryMeta;
+};
+
+type FactoryGenerator<
+  TFactoryData,
+  TVFactoryMeta extends AttachmentFactoryMeta,
+> = Constructable<AttachmentFactory<TFactoryData, TVFactoryMeta>>;
+
+export default class ZipWorkflowManager<
+  TFactoryData,
+  TZipFactoryMeta extends AttachmentFactoryMeta,
+> extends WorkflowManager {
+  protected data: TFactoryData[];
+  protected entityIds: number[] = [];
+  protected factories: AttachmentFactory<TFactoryData, TZipFactoryMeta>[] = [];
+  protected factoriesMeta: Map<number, FactoryMeta<TZipFactoryMeta>> =
+    new Map();
+
+  private responseRS: ReadStream | null = null;
+
+  get responseHeader(): ResponseHeader {
+    return {
+      MIME_TYPE: 'application/zip',
+      CONTENT_DISPOSITION: 'attachment; filename=attachments.zip',
+    };
+  }
+
+  get logPrefix() {
+    return `[${this.constructor.name}] `;
+  }
+
+  constructor(
+    factory: FactoryGenerator<TFactoryData, TZipFactoryMeta>,
+    data: TFactoryData[],
+    extractEntityId: (data: TFactoryData) => number
+  ) {
+    super();
+
+    if (!data) {
+      throw new Error('There is no data to process');
+    }
+    this.data = [...data];
+
+    for (let i = 0; i < this.data.length; i++) {
+      const entityId = extractEntityId(this.data[i]);
+
+      const inst = new factory(entityId);
+
+      this.entityIds.push(entityId);
+
+      inst.onceError((err) => this.ee.emit('error', err));
+      inst.onceDone((meta) => {
+        this.factoriesMeta.set(i, { meta });
+        this.ee.emit('entityExecutionFinished');
+      });
+
+      this.factories.push(inst);
+    }
+
+    this.setupListeners();
+  }
+  start() {
+    for (let i = 0; i < this.data.length; i++) {
+      this.factories[i].init(this.data[i]);
+    }
+  }
+
+  onError(cb: (err: Error) => void): void {
+    this.ee.once('error', (err) => cb(err));
+  }
+
+  onTaskFinished(cb: (rs: Readable) => void): void {
+    this.ee.once('taskFinished', () => {
+      if (this.responseRS === null) {
```

Comment:
I wonder would it be possible that `this.responseRS` is not null instead some other falsy value?

Reply:
We initialise it to null and then set it to a stream that we send back. When an error occurs while setting the stream we handle it differently.